### ISSUE TITLE
Add documentation for JAX foreign function interface

### ISSUE THREAD

**Issue by gnecula** [google/jax#1554]
Document how to add foreign functions written in C++ to JAX, with or without custom gradients, vmap, etc.

Intrernal chat for some background: https://chat.google.com/room/AAAAfoV82X8/bvYqLf3c7Qg

**Comment by gehring** [comment]
@gnecula I can't access the link you posted: "You need a G Suite account to access Hangouts Chat".

**Comment by shoyer** [comment]
@gehring Yes, it's a link to a Google *internal* chat.

The short answer for now: look at `jaxlib/cusolver.*` and `jaxlib/lapack.pyx` for two examples of how to do this.

**Comment by gehring** [comment]
@shoyer  Thanks for the pointers! I figured that was the case. I had assumed the original intent of the link was to give some context to us outsiders which is why I bothered mentioning it. ;)

**Comment by samuela** [comment]
Yes, I have the same question! I posted about this in the slack a while back: https://jaxpy.slack.com/archives/CGR92MPTR/p1581729624022100

**Comment by mattjj** [comment]
@dfm look at this low-numbered issue you get to solve :P

I think #22095 will close this, right?

**Comment by dfm** [comment]
Now that https://github.com/google/jax/pull/22095 is merged, I think this is now documented!

https://jax.readthedocs.io/en/latest/ffi.html
